### ISSUE TITLE
Similar to Puli

### ISSUE THREAD

**Issue by harikt** [imarc/opus#9]
Hey, 

I was looking opus and the idea seems similar to puli : http://docs.puli.io/en/latest/at-a-glance.html

Have you checked it ?

**Comment by mattsah** [comment]
Puli looks nice (at least the docs do anyway).  Opus is a fair bit different, however, in a few ways.  Firstly, it can handle directives for multiple frameworks based on the root package name.  Secondly, it can designate integration packages as copying third-party resources.  Finally, it's aimed at keeping directives maintained in the package itself... this has both pros and cons, but the major pro being that a package can be set to work with multiple frameworks with no additional setup or configuration required by the developer.

In short, it keeps it "packaged."  I don't really see what puli gets me that writing a bash script that I run after composer doesn't.  Also, it's not clear at all if puli does diff checking or integrity checking at all.  Opus directives are designed (although sometimes buggy admittedly) so that you can modify the files it copies and get diffs if they change.

**Comment by harikt** [comment]
@mattsah I don't know more. May be @webmozart can help with the answers ?

**Comment by webmozart** [comment]
Hi :) Thanks for pinging me @harikt, this is really interesting!

I think you're right that Puli is able to do a lot of things that Opus aims to do, although Puli's scope is even broader than that of Opus, which seems to focus about resource copying only (a goal that is going to be handled by a small Puli plugin - see also puli/issues#1).

> it's aimed at keeping directives maintained in the package itself

This is the same for Puli. Each Puli package ships a puli.json file that contains Puli-specific directives. Every package developer has full power over configuring the way their resources are treated, although all configuration can be overridden in the root package so that the application developer has the last word.

> it's not clear at all if puli does diff checking or integrity checking at all

Puli does a lot of conflict detection between packages, but I'm not sure what you mean by "diff checking or integrity checking". Could you elaborate?

> it can designate integration packages as copying third-party resources

Puli does framework integration in a different way than Opus (I think): Frameworks don't really care about the location of resources anymore. Instead, the developer passes a Puli path such as `/imarc/opus/css/style.css`, which the framework delegates to Puli. Puli is able to resolve this path to a filesystem path (e.g. `/path/to/project/vendor/.../css/style.css`) or a path relative to the public directory (e.g. `/opus/css/style.css`). Puli is also able to load resources from other locations than the filesystem (databases, remote servers etc.) so the developer really has a lot of flexibility.

@mattsah I would like to learn more about Opus as I'm sure you put many great ideas into it. Can you tell me more about what you think Opus handles better than Puli?

**Comment by mattsah** [comment]
@webmozart 

Diff and integrity checking is more about conflict detection between the package and the user, and giving the user a simple way to intervene.  If, for example, I have made modifications to a file, and a package has a new version of that file, I'm given the option to see a diff, keep my own, or take theirs.

The point on integration packages was with respect to third party libs that may no provide their own puli.json.  So, for example, if someone maintains a third-party library that I use, but I don't want to maintain a fork of the package itself just to include a puli.json (or in my case Opus directives), I can create an integration package which allows me to say where the assets from those third party packages go from within my own package.

> I would like to learn more about Opus as I'm sure you put many great ideas into it. Can you tell me more about what you think Opus handles better than Puli?

Frankly, I'm not sure I completely understand Puli yet.  This is the first I hear about it, and I only had a chance to skim te docs last night (hence my impression that the puli.json was not something that comes in the package).

That said, I have no idea if Opus does anything better, or really anything much the same.  I'll have more time to look later.

**Comment by mattsah** [comment]
@webmozart 

Am I correct that in order to use Puli I need to use the repositories in my code?

**Comment by webmozart** [comment]
@mattsah Thanks for your response :)

> If, for example, I have made modifications to a file, and a package has a new version of that file, I'm given the option to see a diff, keep my own, or take theirs.

I understand. With Puli, that's not necessary, as you never modify files of other packages. If you want to override a file from another package (say `foo.css` in the package `imarc/opus` which is mapped to the path `/imarc/opus/css/foo.css`), then you do that in your own package by mapping your own file `bar.css` to the same path:

```
puli map /imarc/opus/css/foo.css bar.css
```

This will add two directives to your puli.json:

* The path mapping `/imarc/opus/css/foo.css => bar.css`
* The override statement `your/package => imarc/opus`

This way Puli knows that the file `bar.css` from your package should be used *instead of* the one in `imarc/opus` whenever `/imarc/opus/css/foo.css` is requested.

> I can create an integration package which allows me to say where the assets from those third party packages go from within my own package.

Ah makes sense. With Puli, you can do the same, either directly in your package or in a separate package. For example, you'd use the command:

```
puli map /imarc/opus/css @imarc/opus:assets/css
```

to map the directory `assets/css` in the `imarc/opus` package to the path `/imarc/opus/css`. The mapping is added to the puli.json in the package where you run the command (usually your own root package), so you don't need to touch the `imarc/opus` package at all.

> Am I correct that in order to use Puli I need to use the repositories in my code?

The goal is that average users use an integration layer that does that for them. For example, the Twig extension allows you to use Puli paths in Twig:

```php
echo $twig->render('/imarc/opus/views/index.html.twig');
```

There's no need to deal with the repository directly. I also plan to add CSS pre-processors that replace Puli paths in CSS files by their corresponding web paths and more. You can learn about the currently available extensions on the [Puli Extensions page](http://docs.puli.io/en/latest/extensions.html) in the documentation.

**Comment by mattsah** [comment]
@webmozart 

It's an interesting idea, but certainly different from the principles of Opus.  In particular, I'm not sure that I personally would want to tie my project to any sort of virtual filesystem implementation, but particularly one that works around package dependencies.

Opus is really about giving you pluggable assets or code which are yours to modify, can work with standard PHP and file functions without needing extensions or wrappers, and get version tracked with your project as part of the project.

I think your system probably makes more sense for a CMS plugin arena or something similar where you literally want to be able to include a package that will add functionality to an existing app with a predefined look, style, etc (very little customization to the assets) and I'll definitely be suggesting it for things along those lines.

Opus is much more for the pluggable code at a < app level.  Libraries, frameworks, etc.  Obviously there's some overlap, but I'd hate to be using a system like puli and finding myself having to override almost everything anyway.  Lastly, I think there's a benefit to having things organized on an actual filesystem level a certain way for certain types of code.  If I need to tweak some CSS, I know my starting point is /public/style not... possibly /public/style, but maybe some deeply nested vendor directory somewhere.
